pkg/common/interceptors.go

```diff
@@ -0,0 +1,165 @@
+	log.Infof("GRPC call: %s", info.FullMethod)
+	log.Debugf("GRPC request: %s", protosanitizer.StripSecrets(req))
+	resp, err := handler(ctx, req)
+	if err != nil {
+		log.Errorf("GRPC error: %v", err)
+	} else {
+		log.Debugf("GRPC response: %s", protosanitizer.StripSecrets(resp))
+	}
+	return resp, err
+}
+
+func instrumentGRPC(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler, driverType string) (interface{}, error) {
+	if _, ok := req.(reqWithVolumeId); !ok {
+		return handler(ctx, req)
+	}
+
+	volumeId := req.(reqWithVolumeId).GetVolumeId()
+	method := getMethodName(info.FullMethod)
+	switch method {
+	case ControllerPublish, ControllerUnpublish, NodeStage, NodeUnstage, NodeUnpublish:
+		start := time.Now()
+		resp, err := handler(ctx, req)
+		execTime := time.Since(start).Milliseconds()
+		observeExecTime(execTime, method, driverType, volumeId, err)
+		return resp, err
+	case NodePublish:
+		start := time.Now()
+		resp, err := handler(ctx, req)
+		end := time.Now()
+		execTime := end.Sub(start).Milliseconds()
+		observeExecTime(execTime, method, driverType, volumeId, err)
+		if _, ok := req.(*csi.NodePublishVolumeRequest); !ok {
+			return resp, err
+		}
+		request := req.(*csi.NodePublishVolumeRequest)
+		observeVolumeAttachmentTime(end.UnixMilli(), request, driverType, err)
+		return resp, err
+	}
+	return handler(ctx, req)
+}
+
+func getMethodName(fullName string) string {
+	i := strings.LastIndex(fullName, "/")
+	if !strings.Contains(fullName, "/") || i == len(fullName)-1 {
+		return ""
+	}
+	return fullName[i+1:]
+}
+
+func observeExecTime(time int64, method, driverType, volumeId string, err error) {
+	errCode := getCodeFromError(err).String()
+	metric.CsiGrpcExecTimeCollector.Metric.With(prometheus.Labels{
+		metric.CsiGrpcExecTimeLabelMethod: method,
+		metric.CsiGrpcExecTimeLabelType:   driverType,
+		metric.CsiGrpcExecTimeLabelId:     volumeId,
+		metric.CsiGrpcExecTimeLabelCode:   errCode,
+	}).Observe(float64(time) / 1000)
+}
+
+func getCodeFromError(err error) codes.Code {
+	status, ok := status.FromError(err)
+	if ok {
+		return status.Code()
+	}
+	return codes.Unknown
+}
+
+func observeVolumeAttachmentTime(curTime int64, req *csi.NodePublishVolumeRequest, driverType string, err error) {
+	if client == nil {
+		if client = getK8sClient(); client == nil {
+			log.Warnf("k8s clientset is null, skipping observation for attachment time of volume %s", req.GetVolumeId())
+			return
+		}
+	}
+	errCode := getCodeFromError(err).String()
+	podName, podNamespace := req.VolumeContext[podNameKey], req.VolumeContext[podNamespaceKey]
+	if podName == "" || podNamespace == "" {
+		log.Warnf("observeVolumeAttachmentTime: empty pod name/namespace: %s, %s", podName, podNamespace)
+		return
+	}
+	pod, err := client.CoreV1().Pods(podNamespace).Get(context.Background(), podName, metav1.GetOptions{})
```

Comment:
Additional API server invocation. This can hinder the overall throughput of CSI. We should try to avoid this. Combine it to the runtime class detection?